src/sql/src/query_model/hir.rs

```diff
@@ -0,0 +1,133 @@
+                // @todo self-referencing Maps
+                for scalar in scalars.iter() {
+                    let expr = self.generate_expr(scalar, box_id);
+                    let b = self.model.get_box(box_id);
+                    b.borrow_mut().columns.push(Column { expr, alias: None });
+                }
+                box_id
+            }
+            HirRelationExpr::Project { input, outputs } => {
+                let input_box_id = self.generate_internal(input);
+                let select_id = self.model.make_select_box();
+                let quantifier_id =
+                    self.model
+                        .make_quantifier(QuantifierType::Foreach, input_box_id, select_id);
+                let mut select_box = self.model.get_box(select_id).borrow_mut();
+                for position in outputs {
+                    select_box.columns.push(Column {
+                        expr: BoxScalarExpr::ColumnReference(ColumnReference {
+                            quantifier_id,
+                            position: *position,
+                        }),
+                        alias: None,
+                    });
+                }
+                select_id
+            }
+            _ => panic!("unsupported expression type"),
+        }
+    }
+
+    /// Returns a Select box ranging over the given box, projecting
+    /// all of its columns.
+    fn wrap_within_select(&mut self, box_id: BoxId) -> BoxId {
+        let select_id = self.model.make_select_box();
+        self.model
+            .make_quantifier(QuantifierType::Foreach, box_id, select_id);
+        let mut select_box = self.model.get_box(select_id).borrow_mut();
+        select_box.add_all_input_columns(&self.model);
+        select_id
+    }
+
+    fn generate_expr(&mut self, expr: &HirScalarExpr, context_box: BoxId) -> BoxScalarExpr {
+        match expr {
+            HirScalarExpr::Literal(row, col_type) => {
+                BoxScalarExpr::Literal(row.clone(), col_type.clone())
+            }
+            HirScalarExpr::Column(c) => {
+                assert!(c.level == 0, "correlated columns not yet supported");
+                BoxScalarExpr::ColumnReference(self.find_column_within_box(context_box, c.column))
+            }
+            _ => panic!("unsupported expression type"),
+        }
+    }
+
+    fn find_column_within_box(&self, box_id: BoxId, mut position: usize) -> ColumnReference {
+        let b = self.model.get_box(box_id).borrow();
+        for q_id in b.quantifiers.iter() {
+            let q = self.model.get_quantifier(*q_id).borrow();
+            let ib = self.model.get_box(q.input_box).borrow();
+            if position < ib.columns.len() {
+                return ColumnReference {
+                    quantifier_id: *q_id,
+                    position,
+                };
+            }
+            position -= ib.columns.len();
+        }
+        panic!("column not found")
```

Comment:
Can the panics in this file return the column that was not found? Also, should they be errors since eventually QGM will be used for SQL resolution, so it may be necessary to throw an error in the case of a malformed SQL query.

Reply:
Sure, I'll change this to be an error instead. But this panic is only reached in case of a malformed Hir, which the planner should never produce